### PR TITLE
Users can filter the network, the resources, the forums and events by different fields

### PR DESCRIPTION
**Pull request by Eshah7**
## Trello ticket link

1. [As a user, I want to filter for other user profiles based on characteristics like work experience or interests.](https://trello.com/c/x1k18GhJ/16-as-a-user-i-want-to-filter-for-other-user-profiles-based-on-characteristics-like-work-experience-or-interests)
2. [As a user, I want to be able to filter my resource search results so I can narrow down my desired results.](https://trello.com/c/YSx5gGED/47-as-a-user-i-want-to-be-able-to-filter-my-resource-search-results-so-i-can-narrow-down-my-desired-results)
3. [As a user, I want to be able to filter my forum search results so I can narrow down my desired results.](https://trello.com/c/OalwTeeu/29-as-a-user-i-want-to-be-able-to-filter-my-forum-search-results-so-i-can-narrow-down-my-desired-results)
4.  [As a user, I want to be able to filter my search results for events so that I can narrow down my desired results.](https://trello.com/c/FpVPzI8W/43-as-a-user-i-want-to-be-able-to-filter-my-search-results-for-events-so-that-i-can-narrow-down-my-desired-results)

## Implementation description
* Users can filter the network, the resources, the forums and events by different fields
* Users can clear their filters
* Common components were created to refactor the code

## Steps to test
**Filter Network**
1. Please ensure that you can still search for users by user name
2. Please ensure you can filter by the year and semester
3. Please ensure you can filter by the program
4. Please ensure you can filter by all three, and any two at any time
5. Please ensure you can clear the filter

**Filter Resources**
1. Please select a media tag
2. Ensure that all media with the tag appear 
3. Please ensure that the clear filter feature works

**Filter Forums**
1. Please select a forum tag
2. Ensure that all forums with the tag appear 
3. Please ensure that the clear filter feature works
4. Please edit a forum and ensure that the filtering still works

**Filter Events**
1. Please ensure all of the filters work together
2. Please ensure that the clear filter feature works
3. Please edit an event and ensure that the filter works
4. Please ensure that the events with dates within (including the start and end date) the range appear

## What should reviewers focus on?
* Cards that do not have tags appear with an empty blue circle under the title. 
* The date filter only works when a range is selected, as specified 
* All cards appear, if they meet the filter's requirements

### CODE REVIEW

**Comment by Eshah7** [comment]
[As a user, I want to filter for other user profiles based on characteristics like work experience or interests.](https://trello.com/c/x1k18GhJ/16-as-a-user-i-want-to-filter-for-other-user-profiles-based-on-characteristics-like-work-experience-or-interests)

**Comment by Eshah7** [comment]
[As a user, I want to be able to filter my forum search results so I can narrow down my desired results.](https://trello.com/c/OalwTeeu/29-as-a-user-i-want-to-be-able-to-filter-my-forum-search-results-so-i-can-narrow-down-my-desired-results)

**Comment by Eshah7** [comment]
[As a user, I want to be able to filter my resource search results so I can narrow down my desired results.](https://trello.com/c/YSx5gGED/47-as-a-user-i-want-to-be-able-to-filter-my-resource-search-results-so-i-can-narrow-down-my-desired-results)

**Comment by Eshah7** [comment]
[As a user, I want to be able to filter my search results for events so that I can narrow down my desired results.](https://trello.com/c/FpVPzI8W/43-as-a-user-i-want-to-be-able-to-filter-my-search-results-for-events-so-that-i-can-narrow-down-my-desired-results)

File: src/components/common/filters/ClearFilter.js
Comment (safchow) [inline src/components/common/filters/ClearFilter.js]: should be plural

File: src/components/events/Events.js
Comment (safchow) [inline src/components/events/Events.js:30]: add these filters to the My Events Page

File: src/components/forums/Forums.js
Comment (safchow) [inline src/components/forums/Forums.js:169]: would it be possible to add a date filter, but for when the forum was created?

File: src/components/forums/Forums.js
Comment (safchow) [inline src/components/forums/Forums.js]: add the same filters to the my forums page

File: src/components/resources/CreateResource.js
Comment (safchow) [inline src/components/resources/CreateResource.js]: keep an empty string in the front of the list, the forum uses the first entry in the list as the default value

File: src/components/common/filters/ClearFilter.js
Comment (gabbytang) [inline src/components/common/filters/ClearFilter.js]: I would relabel to "Clear filters", I didn't realize it clears all the filters, not just the one right above the clear filter button

File: src/components/events/Events.js
Comment (gabbytang) [inline src/components/events/Events.js:238]: I think setting the default value of the status filter to be "Active" would be helpful. I think majority of the time, users would want to see the active upcoming events

File: src/components/events/Events.js
Comment (gabbytang) [inline src/components/events/Events.js:247]: Whats the use case for this filter? I'm not sure users will want to search for events based on an exact number of max participants.

What are your thoughts on altering it so that you are searching for events that still have space in the participants list? Ie Are not full

File: src/components/events/Events.js
Comment (gabbytang) [inline src/components/events/Events.js:251]: Is it possible to set the default status to be upcoming events in the next week? Maybe using "TODAY" and setting the end range to be TODAY + week or something? 

Not sure how difficult this is, I know you were having a time with the dumb date issues.

File: src/components/events/Events.js
Comment (gabbytang) [inline src/components/events/Events.js:245]: Can you shrink the font size a teeny bit and make it gray? I feel like it looks a bit off relative to the search bar formatting. Maybe you can make the "Filters" text above it all to be bigger than it

File: src/components/forums/Forums.js
Comment (gabbytang) [inline src/components/forums/Forums.js:158]: Is it possible to make the placeholder text gray?

File: src/components/resources/CreateResource.js
Comment (gabbytang) [inline src/components/resources/CreateResource.js]: I would add an option where a user can pick no tag. Right now the dropdown is preset to "Interview Tips"

File: src/components/resources/Resources.js
Comment (gabbytang) [inline src/components/resources/Resources.js:144]: Can you make this placeholder text match that in the search bar? I think it will make the top of the page look more cohesive. Ie shrink and make gray

File: src/components/resources/Resources.js
Comment (gabbytang) [inline src/components/resources/Resources.js]: What are your thoughts on changing "School Website" to "School". I feel like that's a bit more inclusive and general

**Review by safchow — CHANGES_REQUESTED** [review]

File: src/components/userNetwork/Network.js
Comment (gabbytang) [inline src/components/userNetwork/Network.js:119]: Can you possibly make this placeholder text smaller and gray also? Also is the styling matching the search bar styling, or is Saf covering that later?

**Review by gabbytang — CHANGES_REQUESTED** [review]
Thats crazy that you managed to implement this much stuff already, this is a hella involved user story. This functionality is super cool!!

File: src/components/events/Events.js
Comment (safchow) [reply src/components/events/Events.js:238]: I disagree, the user should be able to view all the events and then filter as needed

File: src/components/events/Events.js
Comment (safchow) [reply src/components/events/Events.js:245]: don't worry about styling, I can do it once its merged in

File: src/components/events/Events.js
Comment (safchow) [reply src/components/events/Events.js:247]: my thought process is that the user is able to remove events from the results of the search that are too large (for example, the user is only looking for events that have 50 or less participants)

File: src/components/events/Events.js
Comment (safchow) [reply src/components/events/Events.js:251]: setting default values for the filters would remove values from the results, and we want to show everything to user first and then allow them to filter from there. I think its better to allow the user to make this decision rather than making it for them if that makes sense

File: src/components/forums/Forums.js
Comment (safchow) [reply src/components/forums/Forums.js:158]: I can style once merged

File: src/components/userNetwork/Network.js
Comment (safchow) [reply src/components/userNetwork/Network.js:119]: I'll cover styling

File: src/components/events/Events.js
Comment (Eshah7) [reply src/components/events/Events.js:251]: The menu already opens to the current month, so I think that should be enough, a default status would cause the filter to enact right away - I think...

File: src/components/resources/CreateResource.js
Comment (Eshah7) [reply src/components/resources/CreateResource.js]: Fixed!

File: src/components/events/Events.js
Comment (gabbytang) [reply src/components/events/Events.js:238]: I disagree with your disagreement. Who wants to see a cancelled event as a default? Why would you want to look a list of events, see one you are really interested in, get excited for it, see participants aren't full, and then notice its cancelled. Why would anyone want to see an event that is no longer happening??? I get that there may be use cases, but I definitely don't go search up and look at a list of events wanting to look at ones that aren't even happening.

File: src/components/events/Events.js
Comment (gabbytang) [reply src/components/events/Events.js:247]: Isn't it looking for exact number? I searched up "24" and nothing showed up, even though most of the events have less then 24 participants as a max/

File: src/components/events/Events.js
Comment (gabbytang) [reply src/components/events/Events.js:251]: I'm thinking its not exactly useful to look at events that have already occurred as a first look at all events.

File: src/components/events/Events.js
Comment (Eshah7) [reply src/components/events/Events.js:238]: i am fine with either - i don't think implementation is too hard but then again, we need to remember that these are filters.

if we want a default look, its better if we render the events like that in the first place IMO

File: src/components/events/Events.js
Comment (Eshah7) [reply src/components/events/Events.js:247]: yeah it is looking for the exact MAX participants but i can try doing the other way!

Comment (safchow) [reply src/components/events/Events.js:247]: you're right, it should be looking for less than the input, not exactly the input. @Eshah7

Comment (Eshah7) [reply src/components/events/Events.js:247]: Updated the feature!

File: src/components/events/Events.js
Comment (gabbytang) [reply src/components/events/Events.js:238]: I still disagree, but I'm outvoted here so I concede

Comment (safchow) [reply src/components/events/Events.js:238]: > I disagree with your disagreement. Who wants to see a cancelled event as a default? Why would you want to look a list of events, see one you are really interested in, get excited for it, see participants aren't full, and then notice its cancelled. Why would anyone want to see an event that is no longer happening??? I get that there may be use cases, but I definitely don't go search up and look at a list of events wanting to look at ones that aren't even happening.

it depends on what we want to show the user on this page. Is this page a list of events both previous and upcoming? Or is it only just a list of events that are upcoming? Ideally, the user is able to make this choice through the filters, and that's why the filter is placed above the list, so that the user makes this decision and then looks at the list. If the user is looking at all the events, we clearly indicate on the card for the event if it's cancelled or not. 

Either way, as long as both cancelled events and active events can show up on this page, I have no issues with showing active events on first render. @gabbytang I'll let you decide since you implemented the events feature :disguised_face:

Comment (gabbytang) [reply src/components/events/Events.js:238]: > > I disagree with your disagreement. Who wants to see a cancelled event as a default? Why would you want to look a list of events, see one you are really interested in, get excited for it, see participants aren't full, and then notice its cancelled. Why would anyone want to see an event that is no longer happening??? I get that there may be use cases, but I definitely don't go search up and look at a list of events wanting to look at ones that aren't even happening.
> 
> it depends on what we want to show the user on this page. Is this page a list of events both previous and upcoming? Or is it only just a list of events that are upcoming? Ideally, the user is able to make this choice through the filters, and that's why the filter is placed above the list, so that the user makes this decision and then looks at the list. If the user is looking at all the events, we clearly indicate on the card for the event if it's cancelled or not.
> 
> Either way, as long as both cancelled events and active events can show up on this page, I have no issues with showing active events on first render. @gabbytang I'll let you decide since you implemented the events feature 🥸

This ain't my pr, I have no hold here other than my opinion. I can see where you both are coming from, though I vehemently disagree with that being the viewpoint in which an end user will hold perspective when using the platform. I 
 personally would expect a "social media" platform to show me upcoming events that I would like to interact and join, having cancelled or passed events be one of the first things I see would not be realistically helpful to me. I would in fact enjoy the option to see cancelled events, but not on initial rendering.

**Review by gabbytang — APPROVED** [review]

**Review by safchow — APPROVED** [review]
lgtm 👍 🤣 💀 🚀

**Pull request merged**
